File: grizon-ai-backend-2-main/Brain/services/command_policy.py
"""WebContainer command policy — block CLI noise, allow npm package installs."""
import re
import shlex
from typing import List, Optional


def is_supabase_cli_command(cmd: str) -> bool:
    c = (cmd or "").strip()
    if not c:
        return False
    if re.match(r"^supabase(\s|$)", c, re.I):
        return True
    if re.search(r"\bsupabase\s+(init|link|login|migration|db|start|stop)\b", c, re.I):
        return True
    return False


def is_echo_instruction(cmd: str) -> bool:
    return bool(re.match(r"^echo\s+", (cmd or "").strip(), re.I))


def is_env_copy_command(cmd: str) -> bool:
    c = (cmd or "").strip().lower()
    return c.startswith("cp .env") or ".env.example" in c


def parse_npm_install_packages(cmd: str) -> Optional[List[str]]:
    m = re.match(r"^npm\s+install\s*(.*)$", (cmd or "").strip(), re.I)
    if not m:
        return None
    tail = (m.group(1) or "").strip()
    if not tail:
        return []
    try:
        tokens = shlex.split(tail)
    except ValueError:
        tokens = tail.split()
    packages = [t for t in tokens if t and not t.startswith("-")]
    return packages


def is_bare_npm_install(cmd: str) -> bool:
    pkgs = parse_npm_install_packages(cmd)
    if pkgs is None:
        return bool(re.match(r"^npm\s+ci\b", (cmd or "").strip(), re.I))
    return len(pkgs) == 0


def should_skip_webcontainer_command(cmd: str) -> bool:
    c = (cmd or "").strip()
    if not c:
        return True
    if is_supabase_cli_command(c):
        return True
    if is_echo_instruction(c):
        return True
    if is_env_copy_command(c):
        return True
    if re.match(r"^npx\s+create[-\s]", c, re.I):
        return True
    # Runner runs bare install + dev/start at the end
    if is_bare_npm_install(c):
        return True
    if re.search(r"\bnpm\s+run\s+dev\b", c, re.I):
        return True
    if re.search(r"\bnpm\s+start\b", c, re.I) or re.search(r"\bnpm\s+run\s+start\b", c, re.I):
        return True
    if re.match(r"^(yarn|pnpm)\s+install\b", c, re.I):
        return True
    return False


def filter_webcontainer_commands(commands: List[str]) -> List[str]:
    out: List[str] = []
    for cmd in commands or []:
        if should_skip_webcontainer_command(cmd):
            print(f"DEBUG: Skipping WebContainer command: {cmd}")
            continue
        out.append(cmd)
    return out


def command_to_op_payloads(cmd: str, cwd: str = "") -> List[dict]:
    """Turn a shell command into structured workspace op payloads (no WorkspaceManager)."""
    c = (cmd or "").strip()
    if not c or should_skip_webcontainer_command(c):
        return []

    packages = parse_npm_install_packages(c)
    if packages is not None and len(packages) > 0:
        op: dict = {"op": "install_packages", "packages": packages}
        if cwd:
            clean = cwd.strip().strip("'\"").lstrip("/").rstrip("/")
            if clean and clean != ".":
                op["cwd"] = clean
        return [op]

    op = {"op": "run", "command": c}
    if cwd:
        clean = cwd.strip().strip("'\"").lstrip("/").rstrip("/")
        if clean and clean != ".":
            op["cwd"] = clean
    return [op]

File: Grizon-AI-Frontend-v2-api-2/brain/lib/commandPolicy.ts
/**
 * WebContainer command policy — allow real npm spawns, block CLI noise only.
 */

export type SpawnSpec = { program: string; args: string[] };

/** Supabase CLI only — not npm packages like @supabase/supabase-js */
export function isSupabaseCliCommand(cmd: string): boolean {
    const c = cmd.trim();
    if (!c) return false;
    if (/^supabase(\s|$)/i.test(c)) return true;
    if (/\bsupabase\s+(init|link|login|migration|db|start|stop)\b/i.test(c)) return true;
    return false;
}

export function isEchoInstruction(cmd: string): boolean {
    return /^echo\s+/i.test(cmd.trim());
}

export function isEnvCopyCommand(cmd: string): boolean {
    const c = cmd.trim().toLowerCase();
    return c.startsWith('cp .env') || c.includes('.env.example');
}

/** True when command should not run in WebContainer (log as skipped instruction). */
export function shouldSkipWebContainerCommand(cmd: string): boolean {
    const c = cmd.trim();
    if (!c) return true;
    if (isSupabaseCliCommand(c)) return true;
    if (isEchoInstruction(c)) return true;
    if (isEnvCopyCommand(c)) return true;
    if (/^npx\s+create[-\s]/i.test(c)) return true;
    return false;
}

/** `npm install` / `npm ci` with no package names (runner runs these at the end). */
export function isBareNpmInstall(cmd: string): boolean {
    const m = cmd.trim().match(/^npm\s+(install|ci)\b/i);
    if (!m) return false;
    const tail = cmd.trim().slice(m[0].length).trim();
    if (!tail) return true;
    const tokens = tail.split(/\s+/);
    const hasPackage = tokens.some((t) => t && !t.startsWith('-'));
    return !hasPackage;
}

export function parseNpmInstallPackages(cmd: string): string[] | null {
    const m = cmd.trim().match(/^npm\s+install\s+(.+)$/i);
    if (!m) return null;
    const tail = m[1].trim();
    if (!tail) return [];
    const packages: string[] = [];
    for (const token of tail.split(/\s+/)) {
        if (!token || token.startsWith('-')) continue;
        packages.push(token);
    }
    return packages;
}

export function devServerKey(cwdKey: string, cmd: string): string {
    const base = cmd.trim().toLowerCase().replace(/\s+/g, ' ');
    return `${cwdKey}::${base}`;
}

/** Prefer spawn(program, args) over jsh -c when we can parse safely. */
export function parseSpawnSpec(cmd: string): SpawnSpec {
    const trimmed = cmd.trim();

    const npmInstall = trimmed.match(/^npm\s+install\s*(.*)$/i);
    if (npmInstall) {
        const tail = (npmInstall[1] || '').trim();
        const args = ['install'];
        if (tail) {
            for (const token of tail.split(/\s+/)) {
                if (token) args.push(token);
            }
        }
        return { program: 'npm', args };
    }

    const npmCi = trimmed.match(/^npm\s+ci\b/i);
    if (npmCi) {
        const tail = trimmed.slice(npmCi[0].length).trim();
        const args = ['ci'];
        if (tail) {
            for (const token of tail.split(/\s+/)) {
                if (token) args.push(token);
            }
        }
        return { program: 'npm', args };
    }

    const npmRun = trimmed.match(/^npm\s+run\s+(\S+)(.*)$/i);
    if (npmRun) {
        const script = npmRun[1];
        const extra = (npmRun[2] || '').trim();
        const args = ['run', script];
        if (extra) {
            for (const token of extra.split(/\s+/)) {
                if (token) args.push(token);
            }
        }
        return { program: 'npm', args };
    }

    if (/^npm\s+start\b/i.test(trimmed)) {
        return { program: 'npm', args: ['start'] };
    }

    return { program: 'jsh', args: ['-c', trimmed] };
}

export function isLongRunningDevCommand(cmd: string): boolean {
    const c = cmd.toLowerCase();
    if (c.includes('install')) return false;
    return (
        c.includes('npm run dev') ||
        c.includes('vite') ||
        c.includes('next dev') ||
        c.includes('npm start') ||
        /node\s+.*server/i.test(c)
    );
}
